text/0000-local-loop-bindings.md

```diff
@@ -0,0 +1,218 @@
+
+An example of a simple iteration, which iterates the loop ten times and prints the iteration number would look like this:
+
+```rust
+loop i = 1 {
+    if i <= 10 {
+        println!("iteration {}", i);
+        i + 1
+    } else {
+        break;
+    }
+}
+```
+
+`continue` will accept an argument in this loop, which will be passed to the next iteration. Using continue, this could look like this:
+
+```rust
+loop i = 1 {
+    if i <= 10 {
+        println!("iteration {}", i);
+        continue i + 1;
+    }
+    break;
+}
+```
+
+Since the end of the loop is never reached, the return value is not required to be the type of the binding, here.
+
+A loop without bindings (`loop { /* body */ }`) will be the same as this:
+
+```rust
+loop () = () {
+    /* body */
+}
+```
+
+This will not be a breaking change, since it's not allowed to have values other than `()` from a loop.
+
+A simple example from the book looks like this:
+
+```rust
+let mut x = 5;
+let mut done = false;
+
+while !done {
+    x += x - 3;
+
+    println!("{}", x);
+
+    if x % 5 == 0 {
+        done = true;
+    }
+}
+```
+
+Using the new syntax, this could be rewritten as this:
+
+```rust
+loop (mut x, done) = (5, false) {
+    if done {
+        break;
+    }
+    x += x - 3;
+
+    println!("{}", x);
+
+    if x % 5 == 0 {
+        (x, true)
+    } else {
+        (x, false)
+    }
+}
+```
+
+This is, how you would define factorial using a loop now:
+
+```rust
+fn factorial(x: i32) -> i32 {
+    loop (result, count) = (1, x) {
+        if count == 1 {
+            break result;
+        }
+        (result * count, count - 1)
+    }
+}
+```
+
+With explicit `continue`, it can look like this:
+
+```rust
+fn factorial(x: i32) -> i32 {
+    loop (result, count) = (1, x) {
+        if count == 1 {
+            break result;
+        } else {
+            continue (result * count, count - 1);
+        }
+    }
+}
+```
+
+Using `break` here allows copying code without having to modify it, when not using a specific function.
+
+Labels will also work. When using `continue` with a label, the arguments to continue must match the loop binding signature connected to the label, in case the label is connected with a loop.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The syntax extension should not cause any issues with backwards compatability.
+
+It's just an extended syntax for `loop` in a place, where currently nothing is allowed yet.
+
+The expansion of the new syntax will be shown for an example.
+
+New syntax:
+
+```rust
+loop (a, mut b) = (x, y) {
+    /* body */
+}
+```
+
+Current syntax:
+
+```rust
+{ // ensure global bindings to be inaccessible after the loop
+    let mut binding = (x, y);
+    loop {
+        let (a, mut b) = binding;
+        binding = {
+            /* body */
+        }
+    }
+}
+```
```

Comment:
An alternative desugaring covering refutable and irrefutable patterns would be:

```rust
loop PAT = EXPR {
    BODY
}

==>

{
    let mut tmp = EXPR;
    loop {
        match tmp {
            PAT => tmp = { BODY },
            _ => break, // If the pattern is irrefutable this will never happen.
        }
    }
}
```

In particular this lets us write:

```rust
loop (mut x, false) = (5, false) {
    x += x - 3;
    println!("{}", x);
    (x, x % 5 == 0)
}
```

Not sure whether this is a good thing, but it seems possible to extend your construct to refutable patterns.

Reply:
Adding this to section "Future possibilities" or maybe "Alternatives" seems better to me.
I'm not sure, if it's a good thing either.
In Scopes it's not, but it also doesn't have inbuilt variant types, so this won't help.